mat64/lq.go

```diff
@@ -9,185 +9,204 @@ import (
 		}
-	} else {
-		for k := 0; k < nh; k++ {
-			hh := f.LQ.RawRowView(k)[k:]
-
-			sub := x.View(k, 0, m-k, n).(*Dense)
-
-			blas64.Gemv(blas.Trans,
-				1, sub.mat, blas64.Vector{Inc: 1, Data: hh},
-				0, blas64.Vector{Inc: 1, Data: proj},
-			)
-			for i := k; i < m; i++ {
-				row := x.RawRowView(i)
-				blas64.Axpy(n, -hh[i-k],
-					blas64.Vector{Inc: 1, Data: proj},
-					blas64.Vector{Inc: 1, Data: row},
-				)
-			}
+		m.mat.Data[i*m.mat.Stride+i] = 1
+		for j := i + 1; j < c; j++ {
+			m.mat.Data[i*m.mat.Stride+j] = 0
 		}
 	}
-}
 
-// Solve computes minimum norm least squares solution of a.x = b where b has as many rows as a.
-// A matrix x is returned that minimizes the two norm of Q*R*X-B. Solve will panic
-// if a is not full rank.
-func (f LQFactor) Solve(b *Dense) (x *Dense) {
-	lq := f.LQ
-	lDiag := f.lDiag
-	m, n := lq.Dims()
-	bm, bn := b.Dims()
-	if bm != m {
-		panic(ErrShape)
+	// Construct Q from the elementary reflectors.
+	h := blas64.General{
+		Rows:   c,
+		Cols:   c,
+		Stride: c,
+		Data:   make([]float64, c*c),
 	}
-	if !f.IsFullRank() {
-		panic(ErrSingular)
+	qCopy := getWorkspace(c, c, false)
+	v := blas64.Vector{
+		Inc:  1,
+		Data: make([]float64, c),
 	}
+	for i := 0; i < r; i++ {
+		// Set h = I.
+		for i := range h.Data {
+			h.Data[i] = 0
+		}
+		for j := 0; j < c; j++ {
+			h.Data[j*c+j] = 1
+		}
 
-	x = NewDense(n, bn, nil)
-	x.Copy(b)
+		// Set the vector data as the elementary reflector.
+		for j := 0; j < i; j++ {
+			v.Data[j] = 0
+		}
+		v.Data[i] = 1
+		for j := i + 1; j < c; j++ {
+			v.Data[j] = lq.lq.mat.Data[i*lq.lq.mat.Stride+j]
+		}
 
-	tau := make([]float64, m)
-	for i := range tau {
-		tau[i] = lq.at(i, i)
-		lq.set(i, i, lDiag[i])
+		// Compute the multiplication matrix.
+		blas64.Ger(-lq.tau[i], v, v, h)
+		qCopy.Copy(m)
+		blas64.Gemm(blas.NoTrans, blas.NoTrans,
+			1, h, qCopy.mat,
+			0, m.mat)
 	}
-	lqT := blas64.Triangular{
-		// N omitted since it is not used by Trsm.
-		Stride: lq.mat.Stride,
-		Data:   lq.mat.Data,
+}
+
+// SolveLQ solves a minimum-norm solution to a system of linear equations defined
+// by the matrices A and B, where A is an m×n matrix represented in its LQ factorized
+// form. If A is singular or near-singular a Condition error is returned. Please
+// see the documentation for Condition for more information.
+//
+// The minimization problem solved depends on the input parameters.
+//  1. If m >= n and trans == false, find X such that || A*X - B||_2 is minimized.
+//  2. If m < n and trans == false, find the minimum norm solution of A * X = B.
+//  3. If m >= n and trans == true, find the minimum norm solution of A^T * X = B.
+//  4. If m < n and trans == true, find X such that || A*X - B||_2 is minimized.
+// The solution matrix, X, is stored in place into the receiver.
+func (m *Dense) SolveLQ(lq *LQ, trans bool, b Matrix) error {
+	r, c := lq.lq.Dims()
+	br, bc := b.Dims()
+
+	// The LQ solve algorithm stores the result in-place into B. The storage
```

Comment:
I meant the command at line 143. The comment is confusing because there is no such thing for in place into B when B is a matrix (lower-case 'm'). If the capital Bs are lower-cased and the capital X is lower-cased I think that fixes it.